DFE-Digital/schools-experience
# Device switching

### JIRA Ticket Number
SE-1961

### Context
Adding subject specific dates to the start of the wizard broke devise switching. After verifying their email the user would be kicked back to chose their subject specific date again as we'd lost the session information.

### Changes proposed in this pull request
Change how browser sessions are mapped to registration_session data. After each step is completed we write the registration_session to redis. The user's browser session now contains a pointer to the registration_session data in redis. We can include this pointer in the magic link emailed to the user, pull it out of the url params when they return, update the pointer in their browser session and when current_registration is next called it will point at the registration_session they left off from.

    ```
    browser_session => {
      'registrations' => {
        'urn' => 'registration_session_uuid'
      }
    }

    redis => {
      'registration_session_uuid' => { registration_session_data }
    }
    ```

### Guidance to review
Manual testing:
Complete registration for a candidate known to gitis but not signed in. Open the verification link in a different browser session, complete the rest of the registration wizard expect not to be kicked back to the beginning of the wizard.
Testing backwards compatibility, over the development branch partially complete the wizard, checkout this branch and carry on with the rest of the wizard, expect not to be kicked back to the beginning.
Cucumber testing, enable the two @wip specs, run the suite, it should pass.
The @wip is there as I suspect that this spec will be flaky when parallel tests are writing to redis as we can't be sure which session we're fishing out.
